### ISSUE TITLE
Taint controller panic error

### ISSUE THREAD

**Issue by gnufied** [kubernetes/kubernetes#51586]
Just saw this while running kube master locally:

```
controller : E0829 23:28:35.313186    4919 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
controller : /data/go/src/runtime/asm_amd64.s:514
controller : /data/go/src/runtime/panic.go:489
controller : /data/go/src/runtime/panic.go:63
controller : /data/go/src/runtime/signal_unix.go:290
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/scheduler/taint_controller.go:287
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/node_controller.go:375
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/util/controller_utils.go:297
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:195
controller : <autogenerated>:57
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:545
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:381
controller : /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
controller : /data/go/src/runtime/asm_amd64.s:2197
controller : panic: runtime error: invalid memory address or nil pointer dereference [recovered]
controller :    panic: runtime error: invalid memory address or nil pointer dereference
controller : [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x1c25ffd]
controller : 
controller : goroutine 953 [running]:
controller : k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
controller : panic(0x2ebb520, 0x4e5cad0)
controller :    /data/go/src/runtime/panic.go:489 +0x2cf
controller : k8s.io/kubernetes/pkg/controller/node/scheduler.(*NoExecuteTaintManager).NodeUpdated(0x0, 0x0, 0xc421902000)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/scheduler/taint_controller.go:287 +0x1ad
controller : k8s.io/kubernetes/pkg/controller/node.NewNodeController.func8(0xc421902000, 0x348f980, 0xc4202a2000)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/node_controller.go:375 +0x44
controller : k8s.io/kubernetes/pkg/controller/node/util.CreateAddNodeHandler.func1(0x348f980, 0xc4202a2000)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/util/controller_utils.go:297 +0x89
controller : k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(0xc4218b5960, 0xc4218b5980, 0xc4218b59a0, 0x348f980, 0xc4202a2000)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:195 +0x49
controller : k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc4218df520, 0x348f980, 0xc4202a2000)
controller :    <autogenerated>:57 +0x73
controller : k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc4218e8370)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:545 +0x287
controller : k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).(k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.run)-fm()
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:381 +0x2a
controller : k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc42025e758, 0xc4218b59d0)
controller :    /home/shared/go_projects/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
controller : created by k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
```

Yet to investigate more closely and check which PR could have introduced this. cc @bowei

**Comment by bowei** [comment]
any way to repro?

**Comment by hzxuzhonghu** [comment]
@gnufied I think I can fix this error.

**Comment by bowei** [comment]
what is the bug?

**Comment by hzxuzhonghu** [comment]
@bowei I think when call `nc.taintManager.NodeUpdated(nil, node)` nc.taintManager has not been inited

**Comment by liggitt** [comment]
duplicate of https://github.com/kubernetes/kubernetes/issues/51013

**Comment by hzxuzhonghu** [comment]
@liggitt I found #51633 does not have node_controller.go related fix.

**Comment by liggitt** [comment]
It fixes the issue systemically by not starting the informers until all controllers have finished setting up

**Comment by hzxuzhonghu** [comment]
Ok ， by more carefully read, I see.

**Comment by lichuqiang** [comment]
@liggitt @deads2k About #51633, some controller's init func(eg. ReplicationController) is overall runs in a goroutine. So it seems  we could not ensure the start order of certain controllers if only the InformersStarted chan is added?

**Comment by liggitt** [comment]
fixed by #51633
